### PR TITLE
Item behaviour revamp :)

### PR DESCRIPTION
**Pull request by Kairus101**
This is now not really the RoS submission, it is a complete reformat of the item system :)
Basically, item behaviours are automatically added through the purchase list - and it handles all the item finding and validating itself. This means bots will have default use for many items without having to add any code into their bot, astrolabe for example. I also extracted out all such items from the bots.

### CODE REVIEW

File: bots/itemBehaviors.lua [hunk 1]
```diff
@@ -0,0 +1,916 @@
...
+	local unitSelf = core.unitSelf
+	return core.OrderItemClamp(botBrain, unitSelf, behaviorLib.itemShrunkenHead)
+end
+behaviorLib.ShrunkenHeadBehavior = {}
+behaviorLib.ShrunkenHeadBehavior["Utility"] = behaviorLib.ShrunkenHeadUtility
+behaviorLib.ShrunkenHeadBehavior["Execute"] = behaviorLib.ShrunkenHeadExecute
+behaviorLib.ShrunkenHeadBehavior["Name"] = "UseShrunkenHead"
+
+------------------------------------
+--  	   Geomancers Bane		  --
```
Comment (mtesija) [inline bots/itemBehaviors.lua]: Is this supposed to be "Geometer's" bane?

Comment (Kairus101) [inline bots/itemBehaviors.lua]: Yes.. :)

File: bots/itemBehaviors.lua [hunk 2]
```diff
@@ -0,0 +1,1044 @@
...
+		local curItem = inventory[slot]
+		if curItem then
+			behaviorLib.addItemBehavior(curItem:GetName(), false)
+		end
+	end
+end
+
+	
+function behaviorLib.addItemBehavior(itemName, remove)
+	local bDebugEchos = true
```
Comment (Anakonda) [inline bots/itemBehaviors.lua]: Turn debugs off.

File: bots/behaviorLib.lua [hunk 3]
```diff
@@ -3724,6 +3299,9 @@ behaviorLib.ShopBehavior["Execute"] = behaviorLib.ShopExecute
 behaviorLib.ShopBehavior["Name"] = "Shop"
 tinsert(behaviorLib.tBehaviors, behaviorLib.ShopBehavior)
 
+--add the item behaviors into the game, so they can be overloaded.
+runfile "bots/itemBehaviors.lua"
```
Comment (Anakonda) [inline bots/behaviorLib.lua]: Are you sure this must be here. I'd like to see it at the top of the file.

File: bots/moonqueen/moonqueen_main.lua [hunk 4]
```diff
@@ -138,6 +138,7 @@ end
 ----------------------------------------------
 -- Find geo, shrunken, rage, helm and boots --
 ----------------------------------------------
+tinsert(behaviorLib.tDontUseDefaultItemBehavior, "Item_ManaBurn2") --don't automatically add default behavior for Geomancers.
```
Comment (Anakonda) [inline bots/moonqueen/moonqueen_main.lua]: Its Geometer's bane :)

File: bots/itemBehaviors.lua [hunk 5]
```diff
@@ -0,0 +1,1044 @@
...
+	-- Ignore items on our ignore list
+	if core.tableContains(behaviorLib.tDontUseDefaultItemBehavior, itemName) > 0 then
+		if bDebugEchos then BotEcho("^rDisabled "..itemName) end
+		return
+	end
+	
+	remove = (remove == nil and false) or remove
+	behaviorLib.behaviorToModify = nil
+	
+	if itemName == "Item_Replenish" then
```
Comment (Anakonda) [inline bots/itemBehaviors.lua]: This is too many elseif in row. Make a table with item name as key and coresponding behavior as value

**Comment by Kairus101** [comment]
Updated :)

**Pull request merged**